contents/handbook/engineering/query-performance-optimization.md

```diff
@@ -1,28 +1,82 @@
+
+1. (if possible) avoid explicit transactions: if you can't, keep them small since transactions lock the processing tables data and may result in deadlocks (super discouraged to use them in application hot paths)
+
+1. (if possible) avoid `JOIN`
+
+1. **always avoid** foreign keys: they are in your way to shard your database, your app is accustomed to rely on them to maintain data integrity instead of doing it on its own, they have performance impact (they add an additional lookup for each insert/delete), they lead to unpredictable performance, ...
+
+1. avoid the use of subqueries: a subquery is a `SELECT` statement that is embedded in a clause of another SQL statement. It's easier to write, but `JOIN`s are usually better-optimized for the database engines.
+
+1. use appropriate [data type(s)](https://www.postgresql.org/docs/10/datatype.html): not all the types occupy the same, and when we use a concrete data type, we can also limit its size according to what we store. For example, `VARCHAR(4000)` is not the same as `VARCHAR(40)`. We always have to adjust to what we will store in our fields not to occupy unnecessary space in our database (and we should enforce this limit in the application code to avoid query errors).
+
+1. use the `LIKE` operator only if necessary: if you know what you are looking for use the `=` operator
+
+#### Indexing
+
+If you are programmatically ordering, sorting, or grouping by a column, you should probably have an index on it. The caveat is that indexing slows down writes to the table and takes disk space (please drop unused indexes).
+
+Composite indices are useful when you want to optimize querying on multiple non-conditional columns. For more info on indices and multi-column indices see the [official docs](https://www.postgresql.org/docs/).
+
+### How-to find slow queries
+
+To find and debug slow queries in production we currently have a single option available: pull real time logs from Heroku by executing `heroku logs --app posthog --ps postgres`
+
+### How-to fix slow queries
+Fixing a slow query is usually a 3 steps process:
+
+1. identify which part of the codebase is generating it (adding the stacktrace as query comments is usually helpful to map query <-> code).
+
+1. re-run the query with `EXPLAIN (ANALYZE, COSTS, VERBOSE, BUFFERS, FORMAT JSON)` as prefix to get the query plan. Query plans aren't the easiest thing to read. They're packed with information and it's closer to being machine parsable than human readable. Postgres Explain Viewer 2 (aka [pev2](https://explain.dalibo.com/), see GitHub [repository](https://github.com/dalibo/pev2)) is a tool to simplify reading query plans. It provides a horizontal tree with each node representing a node in the query plan. It includes timing information, the error amount in the planned versus actual times, and badges for interesting nodes like "costliest" or "bad estimate".
+
+1. fix the query that should now generate a less costly `EXPLAIN` plan.
```

Comment:
non-blocking

For cases that aren't just query changes, e.g. adding indices I guess the story is a bit more involved

Reply:
Yeah, I think we need to iterate more on this document